Modified matrix_ops/tensor_comb.py and matrix_ops/tests/test_tensor_comb.py to enhance tensor_comb

deba-cyber opened the pull request:
## Description
<!--Provide a brief description of the PR's purpose here. If your PR is supposed to fix an existing issue, use
a [keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to link your PR to the issue. -->

## Changes
<!--Notable changes that this PR has either accomplished or will accomplish. Feel free to add more lines to the itemized list
below. -->

  -  [ ] Change 1

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [x] Use `ruff` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [x] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [ ] Use `linkcheck` to check for broken links in the documentation
  -  [ ] Use `doctest` to verify the examples in the function docstrings work as expected.

deba-cyber commented:
Hi @vprusso 
I've changed the tensor_comb and the test for state |ket> or density-matrix. I've tested with pytest on my laptop.
Indentation has been in compliance with pep8. 
Please check it.

vprusso reviewed (commented):
Thanks for the first pass, @deba-cyber. Left some comments. Also, please be sure to run `poetry run ruff check . --fix` to catch and fix any minor linting issues!

vprusso commented:
Hi @deba-cyber looks like there is a conflict and also the comments from the MR are still unresolved. Are you planning on working on this still?

deba-cyber commented:
yes I am working on it @vprusso

vprusso commented:
> yes I am working on it @vprusso

Awesome, thank you for the heads-up, and sorry for the intrusion @deba-cyber !

deba-cyber commented:
Hi @vprusso 
I've modified most of the changes asked. I've tested with ruff and pytest.
I didn't change mode and density_matrix as default values in "tensor_comb.py". 
I felt that user can choose as input whether to have density matrix of ket as output.
I can change it later if still insisted. 
Other things I have incorporated. Should work I believe.
Please check.

vprusso commented:
> I've modified most of the changes asked. I've tested with ruff and pytest.
@deba-cyber it seems like some of the suggestions and comments were not incorporated (but resolved?)

> I didn't change mode and density_matrix as default values in "tensor_comb.py".
I think we should, as commented.

> I felt that user can choose as input whether to have density matrix of ket as output.
They can still choose, but we should provide defaults as suggested.

deba-cyber commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -7,7 +7,9 @@
 
 
-def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+def tensor_comb(k: int, states: list[np.ndarray],
```
Done

deba-cyber commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,44 +1,71 @@
 
-from toqito.matrix_ops import tensor_comb
+from toqito.matrix_ops import tensor_comb, to_density_matrix
 from toqito.states import basis
```
Removed sys

deba-cyber replied:
removed basis

deba-cyber commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +55,51 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
-    possible_sequences = list(itertools.product(range(len(states)), repeat=k))
+    if mode not in ("injective", "non-injective", "diagonal"):
+        raise ValueError(
+            "`mode` must be 'injective','non-injective', or 'diagonal'.")
```
Done.

deba-cyber commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +55,51 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+        case "non-injective":
+            sequences = list(itertools.product(range(len(states)), repeat=k))
+        case "diagonal":
+            sequences = [(i,)*k for i in range(len(states))]
```
Done

deba-cyber commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +55,51 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+
+        if density_matrix:
+            sequences_of_states[seq] = to_density_matrix(
+                sequence_tensor_product)
```
Done

deba-cyber commented:
@vprusso 
Changes suggested have been incorporated. Conflict resolved.

vprusso commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -7,7 +7,11 @@
 
 
-def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+def tensor_comb(
+    k: int,
+    states: list[np.ndarray],
+    mode: str = "non-injective", density_matrix: bool = True,
+    ) -> dict[tuple[int, ...], np.ndarray]:
```
This still seems like it's formatting weird:

```suggestion
def tensor_comb(
    k: int,
    states: list[np.ndarray],
    mode: str = "non-injective",
    density_matrix: bool = True,
) -> dict[tuple[int, ...], np.ndarray]:
```

vprusso commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +57,50 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+        raise ValueError("`mode` must be 'injective','non-injective', or 'diagonal'.")
+
+    if mode == "injective" and k > len(states):
+        raise ValueError("""k must be less than or equal to the number of states for 
+injective sequences.""")
```
```suggestion
        raise ValueError("k must be less than or equal to the number of states for injective sequences.")
```

vprusso commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +57,50 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+        raise ValueError("""k must be less than or equal to the number of states for 
+injective sequences.""")
+
+    # Generate sequences based on the selected mode
```
```suggestion
    # Generate sequences based on the selected mode.
```

vprusso commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +57,50 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+    :return: A dictionary where keys are tuples representing sequences of state indices,
+        and values are density matrices of the tensor products of the corresponding
+        state vectors or tensor products of the corresponding state vectors based on
+        input density_matrix = True / False.
```
```suggestion
        input :code:`density_matrix` being either `True` or `False`.
```

vprusso commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,68 @@
 
 @pytest.mark.parametrize(
-    "states, k, expected",
+    "k, states, mode, density_matrix, expected_keys, expected_error",
```
I don't understand why you removed these test cases. Couldn't you just add another argument for the mode and include the additional test cases here?

vprusso commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,68 @@
-    result = tensor_comb(states, k)
-    for key in expected:
-        assert np.allclose(result[key], expected[key])
+def test_tensor_comb(k, states, mode, density_matrix, expected_keys, expected_error):
+    """Test the tensor_comb function."""
+    if expected_error:
+        with pytest.raises(expected_error):
+            tensor_comb(k, states, mode, density_matrix)
+
+    else:
+        result = tensor_comb(k, states, mode, density_matrix)
+        result_keys = list(result.keys())
+        # check expected keys
+        assert sorted(result.keys()) == sorted(
+            expected_keys)   , f"Expected: {expected_keys}, Got: {list(result.keys())}"
 
+        for seq in expected_keys:
+            state_seq = [states[i] for i in seq]
+            tensor_product = np.array(state_seq[0])
+            for state in state_seq[1:]:
+                tensor_product = np.kron(tensor_product, state)
```
What was wrong with the previous?

```
    result = tensor_comb(states, k)
    for key in expected:
        assert np.allclose(result[key], expected[key])
```

Doing thins like this

```
        # check expected keys
        assert sorted(result.keys()) == sorted(
            expected_keys)   , f"Expected: {expected_keys}, Got: {list(result.keys())}"
```

don't make sense as the pytest framework already does something like this.

vprusso commented:
You may also want to alter the title of your PR to be more descriptive. Right not it's `New branch: changed matrix_ops/tensor_comb.py and matrix_ops/tests/test_tensor_comb.py to add new combinations of tensor product corresponding to enhance tensor_comb #1116` which is quite long and difficult to parse.

deba-cyber commented:
@vprusso 
Changes asked have been incorporated along with change in name of PR. 
Checked through ruff and pytest of course.

vprusso reviewed (commented):
It seems like some of the comments from the prior review were not incorporated, so I flagged those as well. Thanks for the update, and added some additional comemnts.

deba-cyber commented:
@vprusso 
Yes, I missed some of the suggestions before. I've incorporated in this version.

vprusso reviewed (approved):
This LGTM, but I'lll tag @purva-thakre as well to take a look. Thank you, @deba-cyber !

purva-thakre commented:
@deba-cyber You need to resolve some build failures. 

Scroll down and click on one of the failing workflows. 

![image](https://github.com/user-attachments/assets/3a3737eb-9510-4299-8a6a-083df78ec941)

purva-thakre commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +58,48 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
-
+    :param k: The length of the sequence.
+    :param states: A list of state vectors.
+    :mode: Determines the type of sequences.
```
```suggestion
    :param mode: Determines the type of sequences.
```
You can't have a list in a parameter definition. 

![image](https://github.com/user-attachments/assets/9cce3363-2472-4547-bfce-336a790ee6a4)

Check the build locally to verify everything looks fine in your docstring. Instructions are available in the contributing guide. https://toqito--1153.org.readthedocs.build/en/1153/autoapi/matrix_ops/tensor_comb/index.html

purva-thakre commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +58,48 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+    :return: A dictionary where keys are tuples representing sequences of state indices,
+        and values are density matrices of the tensor products of the corresponding
+        state vectors or tensor products of the corresponding state vectors based on
+        input :code:`density_matrix` being either `True` or `False`.
```
```suggestion
        input :code:`density_matrix` being either ``True`` or ``False``.
```

purva-thakre commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,67 @@
+        ([np.array([1, 0]), np.array([0, 1])],
+         2, "invalid", False,
+         None,
+         ValueError),
```
check if the raised error msg matches as expected.

purva-thakre replied:
this one is still unresolved @deba-cyber

Let me know if you can't work on this. I can take over and wrap the PR up.

deba-cyber replied:
@purva-thakre 
Added part for error msg . 
Copied one function for empty states from master branch which I didn't have before.

I didn't want to add another argument in the function as "error msg", hence neede if conditions
to incorporate `invalid' or other case.
If this to be done with match="" kind in the error msg, let me know. But then I believe code needs a bit more
modification.

purva-thakre commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,67 @@
+        ([np.array([1, 0]), np.array([0, 1])],
+         3, "injective", False,
+         None,
+         ValueError),
```
check if the raised error msg matches as expected.

purva-thakre replied:
this one is still unresolved @deba-cyber

deba-cyber replied:
@purva-thakre 
Added part for error msg for invalid.

deba-cyber commented:
@purva-thakre 
New versions of corresponding files have been uploaded. 
"tensor_comb.py" has been modified and "test_tensor_comb.py" has been checked for the error msgs for some test cases.
How to check the build on my local machine , please let me know. 
My OS is Ubuntu 22.04

vprusso commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -53,26 +58,49 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
-        - Keys represent sequences (as tuples) of quantum state indices,
-        - Values are density matrices corresponding to the tensor product of
-          the state vectors for the sequence.
+    :param k: The length of the sequence.
+    :param states: A list of state vectors.
```
Let's order the arguments in the docstring in the same way it is ordered in the prototype. 

```suggestion
    :param states: A list of state vectors.
    :param k: The length of the sequence.    
```

vprusso commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,67 @@
+         2, "non-injective", False,
+         [(0, 0), (0, 1), (1, 0), (1, 1)],
+         None),
+         # Non-injective mode with k = 3
+        ([np.array([1, 0]), np.array([0, 1])],
+         3, "non-injective", False,
+         [(0, 0, 0), (0, 0, 1)],
+         # other possible sequences omitted for brevity
+        None),
```
Shouldn't this test that *all* of the sequences have been generated instead of just the first two?

vprusso commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,67 @@
+            tensor_comb(states, k, mode=mode, density_matrix=density_matrix)
+
+    else:
+        result = tensor_comb(states, k, mode=mode,
+                             density_matrix=density_matrix)
```
```suggestion
        result = tensor_comb(states, k, mode=mode, density_matrix=density_matrix)
```

vprusso commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,67 @@
-    with pytest.raises(ValueError, match="Input list of states cannot be empty."):
-        tensor_comb([], 2)
+            if result[result_keys[0]].ndim == 1:
+                np.testing.assert_allclose(
+                    result[seq], tensor_product, atol=1e-10)
```
```suggestion
                np.testing.assert_allclose(result[seq], tensor_product, atol=1e-10)
```

vprusso commented on toqito/matrix_ops/tests/test_tensor_comb.py:
```diff
@@ -1,50 +1,67 @@
+         2, "invalid", False,
+         None,
+         ValueError),
+        # Injective mode with k = 1
+        ([np.array([1, 0]), np.array([0, 1])],
+         1, "injective", False,
+         [(0,), (1,)],
+         None),
+        # Non-injective mode with k = 1
+        ([np.array([1, 0]), np.array([0, 1])],
+         1, "non-injective", False,
+         [(0,), (1,)],
+         None),
+        # Non-injective mode with k = 2
+        ([np.array([1, 0]), np.array([0, 1])],
+         2, "non-injective", False,
+         [(0, 0), (0, 1), (1, 0), (1, 1)],
+         None),
+         # Non-injective mode with k = 3
+        ([np.array([1, 0]), np.array([0, 1])],
+         3, "non-injective", False,
+         [(0, 0, 0), (0, 0, 1)],
```
```suggestion
        # Injective mode with k = 1.
        ([np.array([1, 0]), np.array([0, 1])],
         1, "injective", False,
         [(0,), (1,)],
         None),
        # Non-injective mode with k = 1.
        ([np.array([1, 0]), np.array([0, 1])],
         1, "non-injective", False,
         [(0,), (1,)],
         None),
        # Non-injective mode with k = 2.
        ([np.array([1, 0]), np.array([0, 1])],
         2, "non-injective", False,
         [(0, 0), (0, 1), (1, 0), (1, 1)],
         None),
         # Non-injective mode with k = 3.
        ([np.array([1, 0]), np.array([0, 1])],
         3, "non-injective", False,
         [(0, 0, 0), (0, 0, 1)],
```

vprusso commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -7,7 +7,12 @@
+    k: int,
+    mode: str = "injective",
+    density_matrix: bool = True
+) -> dict[tuple[int, ...], np.ndarray]:
```
For the examples, can you please use `.. jupyter-execute::` blocks (as is done in, for instance):
https://github.com/vprusso/toqito/blob/e55a29f01efe23c9447f80ee4f42de2e6bdf75f2/toqito/matrices/cyclic_permutation_matrix.py#L16

purva-thakre replied:
Might need to rebase the PR for this.

vprusso reviewed (approved):
Left one minor comment. Also, be sure to respond to the comments from @purva-thakre as well!

purva-thakre commented:
> @purva-thakre New versions of corresponding files have been uploaded. "tensor_comb.py" has been modified and "test_tensor_comb.py" has been checked for the error msgs for some test cases. How to check the build on my local machine , please let me know. My OS is Ubuntu 22.04

Go to the contributing guide in the documentation. Instructions on how to build the documentation locally are already provided.

purva-thakre commented:
@deba-cyber The build failures have not been resolved. See https://github.com/vprusso/toqito/pull/1153#issuecomment-2848229813 and the instructions in the contributing guide.

deba-cyber commented:
@vprusso @purva-thakre 

I used ..jupyter--execute::  in "tensor_comb.py".

I checked the build locally based on documentation. 
For code-style, I ran 
poetry run ruff check .  --fix  
shows all checks passed.

In docs/ 
I ran poetry run sphinx-build -b html . _build
build succeeded, 5 warnings.

While running all tests including pytest, I took screenshot of it, but unable to attach the screenshots for the runs.

I'm not sure about the build failures more than this. Please let me know what I might be missing.

![Screenshot from 2025-05-14 16-14-10](https://github.com/user-attachments/assets/ee32226a-b4da-40be-9c9b-5df5baf486ad)
![Screenshot from 2025-05-14 16-21-35](https://github.com/user-attachments/assets/e7001aa8-5310-4047-990e-e4d43bd24fc6)

deba-cyber commented:
@vprusso @purva-thakre 
![Screenshot from 2025-05-14 16-14-10](https://github.com/user-attachments/assets/8241b799-d54f-42b1-890d-673cd7fbef1c)
![Screenshot from 2025-05-14 16-21-35](https://github.com/user-attachments/assets/a6430a48-ced1-4178-aabb-13fa8f8835ca)
![Screenshot from 2025-05-14 16-23-54](https://github.com/user-attachments/assets/a1488020-16e4-44c1-b16a-173c47129291)
![Screenshot from 2025-05-14 16-37-49](https://github.com/user-attachments/assets/a6c76d8b-468a-4a59-8e92-5e664f880b47)
![Screenshot from 2025-05-14 16-39-51](https://github.com/user-attachments/assets/13678882-3a54-4235-847e-28f871433223)
![Screenshot from 2025-05-14 16-49-54](https://github.com/user-attachments/assets/65339c05-da14-4f7b-b1e4-681f67c0e9fa)

vprusso commented:
@purva-thakre might this have anything to do with the recent doc changes, or is this perhaps something else, anything thoughts?

@deba-cyber also, it looks like there are still some linting errors:
https://github.com/vprusso/toqito/actions/runs/15019404696/job/42218666716?pr=1153

purva-thakre commented:
@vprusso Those are build warnings related to #222 

@deba-cyber `ruff` is a linter and a formatter. `code-style` is failing due to formatting issues.

purva-thakre commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -30,42 +35,66 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
-          the state vectors for the sequence.
+    :param states: A list of state vectors.
+    :param k: The length of the sequence.
+    :param mode: Determines the type of sequences.
+            - "non-injective": Allows repetitions in sequences.
+            - "injective": Ensures sequences are injective (no repetitions).
+            - "diagonal": Only sequences with repeated indices (diagonal elements).
```
You cannot add a list to a `param`. As I said before, check the build locally to verify the documentation appears as expected. 

![image](https://github.com/user-attachments/assets/3ae5dd54-2229-4a22-9a97-6b1934a76f8f)

purva-thakre commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -30,42 +35,66 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+    :param states: A list of state vectors.
+    :param k: The length of the sequence.
+    :param mode: Determines the type of sequences.
+            - "non-injective": Allows repetitions in sequences.
```
```suggestion
            -`non-injective`: Allows repetitions in sequences.
```
Apply the same to `injective` and `diagonal`.

deba-cyber commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -30,42 +35,66 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
+            - "non-injective": Allows repetitions in sequences.
+            - "injective": Ensures sequences are injective (no repetitions).
+            - "diagonal": Only sequences with repeated indices (diagonal elements).
+    :density_matrix: Determines whether the return is density matrix or ket.
```
@purva-thakre 
done

purva-thakre commented on toqito/matrix_ops/tensor_comb.py:
```diff
@@ -30,42 +35,64 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
-          the state vectors for the sequence.
+    :param states: A list of state vectors.
+    :param k: The length of the sequence.
+    :param mode: Determines the type of sequences.-`non-injective`: Allows repetitions in
+           sequences.-`injective`: Ensures sequences are injective (no repetitions).-`diagonal`: Only
+           sequences with repeated indices (diagonal elements).
```
Previous comment about this is still unresolved. You cannot add a list here.

If you are using `-`, this is an attempt at creating a list.

purva-thakre commented:
@deba-cyber Out of curioisity, how are you adding commits to this PR?

Each commit is along the lines of `Add files via upload`.

purva-thakre commented:
Thank you for your help @deba-cyber

I have taken your commits into #1219
